module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+            }
+        };
+
+        // This proof checks that the attester was bonded at the given transition num
+        let bond_opt = working_set
+            .backing()
+            .verify_proof(
+                bonding_root,
+                attestation.proof_of_bond.proof.clone(),
+                context.sender(),
+                &self.bonded_attesters,
+            )
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondingProof)?;
+
+        let bond = bond_opt.unwrap_or_default();
+        let bond: u64 = BorshDeserialize::deserialize(&mut bond.value())
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondFormat)?;
+
+        let minimum_bond = self
+            .minimum_attester_bond
+            .get_or_err(working_set)
+            .expect("The minimum bond should be set at genesis");
+
+        // We then have to check that the bond was greater than the minimum bond
+        if bond < minimum_bond {
+            return Err(AttesterIncentiveErrors::UserNotBonded);
+        }
+
+        Ok(())
+    }
+
+    fn check_transition(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        if let Some(curr_tx) = self
+            .chain_state
+            .historical_transitions
+            .get(&max_attested_height, working_set)
+        {
+            // We first need to compare the initial block hash to the previous post state root
+            if !curr_tx.compare_hashes(&attestation.da_block_hash, &attestation.post_state_root) {
+                // It is the right attestation, we have to compare the initial block hash to the
+                // previous post state root
+                // Slash the attester
+                return Err(self.slash_and_invalidate_attestation(
+                    attester,
+                    max_attested_height,
+                    SlashingReason::TransitionInvalid,
+                    working_set,
+                ));
+            }
+            Ok(CallResponse::default())
+        } else {
+            Err(self.slash_burn_reward(
+                attester,
+                Role::Attester,
+                SlashingReason::TransitionNotFound,
+                working_set,
+            ))
+        }
+    }
+
+    fn check_initial_hash(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // Normal state
+        if let Some(transition) = self
+            .chain_state
+            .historical_transitions
+            .get(&(max_attested_height - 1), working_set)
```

Comment:
Can this underflow? It looks like it to me - but let's use `saturating_sub` here in any case so that it's easier to tell at a glance.